test/test_torch.py

```diff
@@ -5302,6 +5297,138 @@ def add_neg_dim_tests():
 class TestTorchDeviceType(TestCase):
     exact_dtype = True
 
+    def test_assertEqual_debug(self, device):
+        # Scalar comparison
+        s = ".+Comparing 4.0 and 7.0 gives a difference of 3.0.+"
+        with self.assertRaisesRegex(AssertionError, s):
+            self.assertEqual(4., 7.)
+
+        # Scalar complex comparison, real part
+        s = ".+Comparing the real part 1.0 and 3.0 gives a difference of 2.0.+"
+        with self.assertRaisesRegex(AssertionError, s):
+            self.assertEqual(complex(1, 3), complex(3, 1))
+
+        # Scalar complex comparison, imaginary part
+        s = ".+Comparing the imaginary part 3.0 and 5.5 gives a difference of 2.5.+"
+        with self.assertRaisesRegex(AssertionError, s):
+            self.assertEqual(complex(1, 3), complex(1, 5.5))
+
+    @unittest.skipIf(not TEST_NUMPY, 'NumPy not found')
+    @dtypes(*list(product(torch_to_numpy_dtype_dict.keys(),
+                          torch_to_numpy_dtype_dict.keys())))
+    def test_compare_type(self, device, dtypes):
+        a = torch.empty(0, device=device, dtype=dtypes[0])
+        b = torch.empty(0, device=device, dtype=dtypes[1])
+        torch_compare_dtype = self.compare_type(a, b, allow_ints=True)
+        np_compare_dtype = np.promote_types(
+            torch_to_numpy_dtype_dict[dtypes[0]],
+            torch_to_numpy_dtype_dict[dtypes[1]])
+
+        # Note: NumPy returns instances of dtypes so this converts them
+        # to their more commonly used representation
+        # (ex. dtype('int16') -> np.int16)
+        np_compare_dtype = getattr(np, str(np_compare_dtype))
+
+        # Special-cases float16 inputs on non-CUDA devices
+        if np_compare_dtype is np.float16 and self.device_type != 'cuda':
+            self.assertTrue(torch_compare_dtype is torch.float32)
+            return
+
+        # Checks that float16 and bfloat16 have the same promotion on CPU
+        if self.device_type == 'cpu' and dtypes[0] is torch.float16:
+            c = torch.empty(0, device=device, dtype=torch.bfloat16)
+            bfloat16_type = self.compare_type(c, b, allow_ints=True)
+            self.assertTrue(torch_compare_dtype is bfloat16_type)
+
+        self.assertTrue(torch_compare_dtype ==
+                        numpy_to_torch_dtype_dict[np_compare_dtype])
+
+    # Checks that compareTensors provides the correct debug info
+    def test_comparetensors_debug(self, device):
+        # Checks float tensor comparisons (2D tensor)
+        a = torch.tensor(((0, 6), (7, 9)), device=device, dtype=torch.float32)
+        b = torch.tensor(((0, 7), (7, 22)), device=device, dtype=torch.float32)
+        result, debug_msg = self.compareTensors(a, b)
+        expected_msg = ("With rtol=1.3e-06 and atol=1e-05, found 2 element(s) "
```

Comment:
Instead of hardcoding `rtol/atol`, is there any way that these can be passed in? Since they are overrideable by backends. Thanks!

Reply:
This change maintains XLA's ability to override atol by setting the "precision" argument on the class ;)

Reply:
Oh I see what you mean now - the strings are wrong on XLA. Duh. Yeah I'll fix this.

Reply:
Unfortunately even after updating the message text it looks like XLA still can't pass the test because it can't convert complex->float. For example, on Colab:

```
t = torch.tensor((1 + 1j, 2 + 2j), device=dev).to(torch.float32)
t
```

will throw the error:

```
RuntimeError                              Traceback (most recent call last)
/usr/local/lib/python3.6/dist-packages/IPython/core/formatters.py in __call__(self, obj)
    697                 type_pprinters=self.type_printers,
    698                 deferred_pprinters=self.deferred_printers)
--> 699             printer.pretty(obj)
    700             printer.flush()
    701             return stream.getvalue()

6 frames
/usr/local/lib/python3.6/dist-packages/torch/_tensor_str.py in __init__(self, tensor)
     86 
     87         else:
---> 88             nonzero_finite_vals = torch.masked_select(tensor_view, torch.isfinite(tensor_view) & tensor_view.ne(0))
     89 
     90             if nonzero_finite_vals.numel() == 0:

RuntimeError: Error while lowering: f32[2]{0} xla::cast, type=f32, dtype=Float, stype=ComplexFloat
Error: torch_xla/csrc/convert_ops.cpp:83 : Unsupported XLA type 15
Python Frames:
```

Reply:
Hmmm I guess it might work with https://github.com/pytorch/xla/pull/2003 since we convert tensor to cpu before doing comparison. But yeah skipping is also fine. :D 
Btw would it make sense to make it `_compareTensors` to say it's an internal API so that people still use `assertEqual`?

Reply:
Sure. Sounds good!